Feature/#336 restrict challenge creation to admins

shernandez334 opened the pull request:
Restringir el acceso al endpoint addChallenge() para que únicamente usuarios autenticados y autorizados (por ejemplo, con rol de administrador o creador) puedan crear nuevos retos.

Actualmente, este endpoint está expuesto y permite a cualquier usuario realizar esta acción. Se requiere implementar un mecanismo de control de acceso (por ejemplo, con Spring Security) para garantizar que solo ciertos perfiles puedan acceder a esta funcionalidad.

La securizacion sera basada en el endpoint addChallengeToFavorite().

Commit:
feature/#336-restrict-challenge-creation-to-admins

# Conflicts:
#	itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java
#	itachallenge-challenge/src/main/java/com/itachallenge/challenge/exception/GlobalExceptionHandler.java

# Conflicts:
#	itachallenge-auth/src/main/java/com/itachallenge/auth/controller/AuthController.java
#	itachallenge-auth/src/main/java/com/itachallenge/auth/service/JwtService.java

Commit:
Actualización del workflow feature/#336-restrict-challenge-creation-to-admins: cambio de versión de Gradle

Commit:
Actualización del workflow test feature/#336-restrict-challenge-creation-to-admins: cambio de versión de Gradle

Commit:
Enable full warnings and stacktrace in test step

Commit:
Revert deploy and request_pr workflows to develop branch versions and fix addChallenge Tests

Commit:
refactor(jwt): extract 'Bearer ' prefix into constant to remove duplication

Replaced multiple instances of the hardcoded 'Bearer ' string in JwtServiceImpl with a single constant (BEARER_PREFIX). This improves maintainability, reduces duplication, and resolves a SonarCloud code smell warning about repeated literals. No logic was changed; only internal structure was improved.

Commit:
Removed authExceptionHandler

Commit:
Added test for SonarCloud

Commit:
Added test for SonarCloud 2

Commit:
Remove unnecesary exceptions in the folder auth

Commit:
Undo file changes

Commit:
removing unnecesary changes

Commit:
Adding necessary test for challenge

trisk910 commented on itachallenge-auth/src/main/java/com/itachallenge/auth/controller/AuthController.java:
Habría que restringir el acceso a este endpoint o crear un rol/perfil que sea dev:  @Profile("dev"). Para evitar accesos a los endpoints.

ivilarop commented on itachallenge-auth/src/main/java/com/itachallenge/auth/controller/AuthController.java:
```diff
@@ -159,6 +159,4 @@ public Mono<ResponseEntity<Map<String, String>>> logout(@RequestHeader (value =
                     .body(Map.of("message", "Invalid or expired token")));
         }
     }
-
```
If this class is not modified, it should not actually appear as a change. Traceability is then lost at the historical level.

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java:
```diff
@@ -217,9 +218,18 @@ public Mono<Map<String, Object>> addSolution(@Valid @RequestBody SolutionDto sol
     )
-    public Mono<ResponseEntity<ChallengeDto>> addChallenge(@Valid @RequestBody ChallengeCreateDto createFormDto) {
+
+    public Mono<ResponseEntity<ChallengeDto>> addChallenge(
```
Currently, only mentors can log in to the project, so any valid JWT token already implies that the user has permission to perform this action.

Therefore, in the remaining endpoints (such as addChallengeToFavorite, removeChallengeFromFavorite, etc.), we simply extract the userId from the token using jwtService.getUserUuIdFromAuthenticationHeader(...), without validating the role.

For consistency and simplicity, it would be more appropriate to apply the same pattern here and remove the isAdmin(...) validation, since we're not currently working with different roles.

🔎 Additionally, this is exactly what was requested in user story #321: securing the endpoint with the JWT filter, as already implemented in other methods, without adding additional role logic.

ivilarop commented:
@shernandez334 changelog missing

Commit:
refactor: remove redundant role validation from challenge endpoints

# Conflicts:
#	itachallenge-auth/src/main/java/com/itachallenge/auth/controller/AuthController.java

Commit:
remove unused dependencies

Commit:
remove unnecessary tests

Commit:
Fixed ChallengeController Tests

shernandez334 committed:
Update application.yml

shernandez334 committed:
Update application.yml

shernandez334 commented:
> @shernandez334 changelog missing

Changelog added.

Commit:
reverted back to develop branch AuthController

Commit:
Solve branch conflicts

IsmaPeiro commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java:
```diff
@@ -312,6 +317,7 @@ public Mono<ResponseEntity<FavoriteDto>> removeChallengeFromFavorite(
             responses = {
                     @ApiResponse(responseCode = "200", content = {@Content(schema = @Schema(implementation = FavoriteDto.class), mediaType = "application/json")}),
                     @ApiResponse(responseCode = "400", description = "Missing or invalid authorization header."),
+                    @ApiResponse(responseCode = "400", description = "Missing or invalid authorization header."),
```
I noticed there are two identical
@ApiResponse(responseCode = "400", description = "Missing or invalid authorization header.")
annotations. To avoid duplication, you should remove one of them.

IsmaPeiro replied:
I noticed there are two identical
@ApiResponse(responseCode = "400", description = "Missing or invalid authorization header.")
annotations. To avoid duplication, you should remove one of them.

IsmaPeiro replied:
I noticed there are two identical
@ApiResponse(responseCode = "400", description = "Missing or invalid authorization header.")
annotations. To avoid duplication, you should remove one of them.

IsmaPeiro commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java:
```diff
@@ -346,6 +352,7 @@ public Mono<GenericResultDto<TagDto>> getAllTags() {
             description = "Allows to update any information contained in a challenge, providing ChallengeId and new information.",
             responses = {
                     @ApiResponse(responseCode = "200", content = {@Content(schema = @Schema(implementation = ChallengeDto.class), mediaType = "application/json")}),
+                    @ApiResponse(responseCode = "400", description = "Missing or invalid authorization header."),
```
@ApiResponse(responseCode = "400", description = "Missing or invalid authorization header.") – I don’t think this should be here since there’s currently no header validation or usage. It would be better to use something like
@ApiResponse(responseCode = "400", description = "Missing or invalid request body.")

IsmaPeiro replied:
@ApiResponse(responseCode = "400", description = "Missing or invalid authorization header.") – I don’t think this should be here since there’s currently no header validation or usage. It would be better to use something like
@ApiResponse(responseCode = "400", description = "Missing or invalid request body.")

IsmaPeiro commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/dto/ChallengeDto.java:
```diff
@@ -61,7 +61,7 @@ public class ChallengeDto {
     private List<UUID> tags;
 
-    @JsonProperty(index = 12)
+    @JsonProperty(index = 11)
```
You’ve updated the timesBookmark index from 12 to 11 to avoid a conflict with the timesSolved index. However, the value 11 was already assigned to tags, creating a new conflict.

IsmaPeiro replied:
I think you’ve updated the timesBookmark index from 12 to 11 to avoid a conflict with the timesSolved index. However, the value 11 was already assigned to tags, creating a new conflict.

IsmaPeiro reviewed (commented):
Hello, I’ve left some comments in the code—let me know what you think. Besides that, I noticed there are some changes that are just line breaks being removed and then re-added, which don’t actually change anything in the code.

Commit:
Removed duplicated code in ChallengeController

shernandez334 committed:
Update TagServiceImpl.java

shernandez334 committed:
Update TagServiceImpl.java

shernandez334 committed:
Update GlobalExceptionHandler.java

shernandez334 committed:
Update TagDocument.java

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/controller/ChallengeController.java:
```diff
@@ -219,9 +219,15 @@ public Mono<Map<String, Object>> addSolution(@Valid @RequestBody SolutionDto sol
+            @RequestHeader(name = "Authorization", required = false) String authHeader) {
+
+        return Mono.fromCallable(() -> jwtService.getUserUuIdFromAuthenticationHeader(authHeader))
+                .flatMap(userId -> challengeService.addChallenge(createFormDto))
```
And the onErrorMap error? Why haven't you followed the other endpoints already implemented that use it?

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/service/JwtServiceImpl.java:
```diff
@@ -14,13 +14,14 @@
 public class JwtServiceImpl implements IJwtService {
 
     private static final Logger log = LoggerFactory.getLogger(JwtServiceImpl.class);
+    private static final String BEARER_PREFIX = "Bearer ";
```
These changes are also made in the PR https://github.com/IT-Academy-BCN/ita-challenges-backend/pull/875. Why are they duplicated?

ivilarop replied:
Also check out the comments I made regarding these changes in the other PR

ivilarop commented on itachallenge-challenge/src/main/java/com/itachallenge/challenge/exception/GlobalExceptionHandler.java:
```diff
@@ -122,4 +122,4 @@ private Optional<ResponseEntity<MessageDto>> buildTagUuidError(InvalidFormatExce
         }
         return Optional.empty();
     }
-}
```
Why did we upload this class? We're cluttering Git's history with classes that aren't actually changing.

trisk910 committed:
Update GlobalExceptionHandler.java

Commit:
Adjusted the addChallenge to follow the other endpoints patterns

trisk910 committed:
Update GlobalExceptionHandler.java

Commit:
Changes

Commit:
Fixed conflict issues

shernandez334 committed:
Update TagServiceImpl.java
